compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs

```diff
@@ -158,13 +158,14 @@ onExp (L o (SectionR _ mid@(isDot -> True) rhs))
     , srcSpanStart o == srcSpanStart (getLoc mid)
     , srcSpanEnd o == srcSpanEnd (getLoc rhs)
     , Just sels <- getSelectors rhs
-    = setL o $ mkParen $ foldl1 (\x y -> noL $ OpApp noE x (mkVar var_dot) y) $ map (mkVar var_getField `mkAppType`) $ reverse sels
+    -- Don't bracket : it came in as a section so it's already enclosed.
+    = setL o $ foldl1 (\x y -> noL $ OpApp noE x (mkVar var_dot) y) $ map (mkVar var_getField `mkAppType`) $ reverse sels
 
 -- Turn a{b=c, ...} into setField calls
 onExp (L o upd@RecordUpd{rupd_expr,rupd_flds=L _ (HsRecField (fmap rdrNameAmbiguousFieldOcc -> lbl) arg pun):flds})
     | let sel = mkSelector lbl
     , let arg2 = if pun then noL $ HsVar noE lbl else arg
-    , let expr = mkParen $ mkVar var_setField `mkAppType` sel `mkApp` arg2 `mkApp` mkParen rupd_expr
+    , let expr = mkParen $ mkVar var_setField `mkAppType` sel `mkApp` arg2 `mkApp` rupd_expr -- 'rupd_expr' never appears to need bracketing.
```

Comment:
“never appears to need bracketing” seems a bit too unsure to make me comfortable :slightly_smiling_face:. I don’t really know how GHC resolves precedences but it seems like you could end up with `(setField @"foo" x y) + z` instead of `setField @"foo" x (y + z)`. Have you tested this and similar cases, e.g., it would also be interesting to have another function application there so `y z`.

Reply:
We probably at least want to have some test cases that hit this.

Reply:
`rupd_expr` is the record part of the update in `RecordUpd {rupd_expr; rupd_fields}`. I think there are 3 possibilities : `rupd_expr` is a var, `rupd_expr` is an application, `rupd_expr` is a record constructor/update. (1) doesn't need bracketing, (2) needs to be in brackets already or it won't parse and (3) never needs bracketing.

Reply:
I'll fix the comment. I agree it doesn't inspire confidence :) I wrote it based on practice before I'd convinced myself by theory.

Reply:
(1) `-- 'rupd_expr' never needs bracketing.`
(2) ``` -- Don't bracket here. The argument came in as a section so it's
    -- already enclosed in brackets.```

Reply:
I’m still not convinced, sorry :slightly_smiling_face:  In particular, I am not convinced that 2 holds. If I have `x { a = y, b = z }` I believe you will first rewrite this to `setField @"a" x { b = z }` and then when you rewrite this it breaks.

Reply:
In the above `x` is the `rupd_expr`.

Reply:
Test:
```
data C = C {a : Int, b: Int}
g x y z = x {a = y, b = z}
```
Compiles and is lint free.

Reply:
Which 2 is it that you are not convinced holds?

Reply:
> “never appears to need bracketing” seems a bit too unsure to make me comfortable 🙂. I don’t really know how GHC resolves precedences but it seems like you could end up with `(setField @"foo" x y) + z` instead of `setField @"foo" x (y + z)`. Have you tested this and similar cases, e.g., it would also be interesting to have another function application there so `y z`.

Here's some tests that I think covers those?
```
data C = C {a : Int, b: Int}
g x y z = x {a = y, b = z}
h x y z = x {a = y + z}

mkC a b = C with a, b

main = scenario do
  let x = C {a = 1, b = 2}
  let x' = g x 3 4
  assert $ x'.a == 3 && x'.b == 4

  let x'' = h x 1 2
  assert $ x''.a == 3

  let x''' = h (mkC 5 6) 1 1
  assert $ x'''.a == 2
```

Reply:
One more from above you asked for and I missed `i x y =  x {a = succ y}` :
```
  let x'''' = i C {a = 1, b = 2} 2
  assert $ x''''.a == 3
```

Reply:
If it works then all is good but I’m still not sure I understand how it works. Is it as simple as GHC doesn’t care if `rupd_expr` has parentheses or not? It seems like you will definitely not always have parentheses around the expression if there are multiple field updates.